distributed/protocol/pickle.py

```diff
@@ -1,19 +1,36 @@
 from __future__ import annotations
 
 import inspect
+import io
 import logging
 import pickle
 
 import cloudpickle
 from packaging.version import parse as parse_version
 
+from distributed.protocol.serialize import dask_deserialize, dask_serialize
+
 CLOUDPICKLE_GTE_20 = parse_version(cloudpickle.__version__) >= parse_version("2.0.0")
 
 HIGHEST_PROTOCOL = pickle.HIGHEST_PROTOCOL
 
 logger = logging.getLogger(__name__)
 
 
+class _DaskPickler(pickle.Pickler):
+    def reducer_override(self, obj):
+        # For some objects this causes segfaults otherwise, see
+        # https://github.com/dask/distributed/pull/7564#issuecomment-1438727339
+        if _always_use_pickle_for(obj):
+            return NotImplemented
+        try:
+            serialize = dask_serialize.dispatch(type(obj))
+            deserialize = dask_deserialize.dispatch(type(obj))
+            return deserialize, serialize(obj)
+        except TypeError:
```

Comment:
I assume we should be dealing with other families, e.g. cuda as well here, don't we? any suggestions on how to do this in an elegant way?

Reply:
A possible way of dealing with this would be a custom dispatcher that controls this specific behavior and isn't entangled with the other serialization logic. Thoughts?